### DISCUSSION TITLE
Permissions in the CI/CD Pipeline

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#491] [Help] [answered]
How do I control who can do what in the CI/CD pipeline?

For example, suppose I want to allow some developers to deploy any infrastructure they want in the `dev` account, but in the `stage` account, I want to enforce peer approval of a PR, and in the `prod` account, I want only a select group of people to be able to approve and merge the PR.

Does this require breaking each of the accounts into separate Github repositories, so that each repository can have its own Github workflow and CircleCI (in our case) configuration?

Do you have examples of this in action?

---

<ins datetime="2022-07-06T19:51:55Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108931">Tracked in ticket #108931</a></p>
</ins>

**Comment by yorinasub17** [comment, marked as answer]
In general, it is very hard to restrict the CI/CD flow when someone has write access to a repository. While the various VCS systems have some form of branch protection feature, these features are generally fairly limited in what you can limit.

In most cases, if you have advanced use cases with ACLs and CI/CD, it is best to break apart the code for each of the environment subsets into separate repos so that you only grant write access to the users that are allowed to freely deploy.

Note that depending on the VCS system, you may be able to craft a branch protection rule that allows implementation of the flow you would like. For example, on GitHub, you can implement branch protection with `CODEOWNERS` such that you must have a PR to merge into `main`, and every PR requires an approval from a `CODEOWNER`. Then, you can implement code owner rules such that `prod` and `stage` have a different set of users who can approve the code to deploy.

Be aware that even this has limitation, as you can't dynamically adjust the number of reviews that are required. So you wouldn't be able to specify that changes to the `dev` folder are allowed to be merged without approval.

**Reply by pdeva** [reply]
given the above, why is it then that the ‘reference architecture’ from gruntwork has a monorepo with terragrunt configurations for each environment in the same repo. wouldn't an ‘environment per repo’ setup be better suited to easily enforce proper permissions for deploying in each environment?

**Reply by pete0emerson** [reply]
To address the specific issue of advanced uses with ACLs and CI/CD, a polyrepo approach may be better for you, but it comes with trade-offs. For example, our approach allows for the elimination of having the same code in multiple places (the `_envcommon` structure, for example). The points in our [terragrunt-infrastructure-modules-example](https://github.com/gruntwork-io/terragrunt-infrastructure-modules-example#monorepo-vs-polyrepo) go into more detail about the pros and cons.

From our perspective it doesn't make sense for us to deliver two types of Reference Architectures (one monorepo, one polyrepo). We've decided that the monorepo approach works best for us, and hopefully it works for most of our customers, as well.
